rayon-core/src/latch.rs

```diff
@@ -374,7 +383,7 @@ where
     L: Latch,
 {
     #[inline]
-    fn set(&self) {
-        L::set(self);
+    unsafe fn set(this: *const Self) {
+        L::set(&**this);
     }
```

Comment:
This impl makes me uncomfortable. I previously tried to implement this PR and I think I came up with almost exactly the same patch as you have here, but I wasn't sure what to do with this. Of course I couldn't come up with a case where this actually causes an issue, but implementing this for `&Latch` seems like an invitation to create the references we're trying to avoid.

Reply:
I added a new `LatchRef` to replace this -- let me know what you think!

Reply:
That is very cool. I like it.